kafka/src/main/java/com/metamx/tranquility/kafka/KafkaConsumer.java

```diff
@@ -0,0 +1,273 @@
+    this.commitThread.setName("KafkaConsumer-CommitThread");
+    this.commitThread.setDaemon(true);
+  }
+
+  public void start()
+  {
+    commitThread.start();
+    startConsumers();
+  }
+
+  public void stop()
+  {
+    if (shutdown.compareAndSet(false, true)) {
+      log.info("Shutting down - attempting to flush buffers and commit final offsets");
+
+      try {
+        commitLock.writeLock().lockInterruptibly(); // prevent Kafka from consuming any more events
+        try {
+          writerController.flushAll(); // try to flush the remaining events to Druid
+          writerController.stop();
+          consumerConnector.commitOffsets(); // update commit offset
+        }
+        finally {
+          commitLock.writeLock().unlock();
+          consumerConnector.shutdown();
+          commitThread.interrupt();
+          consumerExec.shutdownNow();
+        }
+      }
+      catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        Throwables.propagate(e);
+      }
+
+      log.info("Finished clean shutdown.");
+    }
+  }
+
+  public void join() throws InterruptedException
+  {
+    commitThread.join();
+  }
+
+  void commit() throws InterruptedException
+  {
+    commitLock.writeLock().lockInterruptibly();
+    try {
+      final long flushStartTime = System.currentTimeMillis();
+      final Map<String, MessageCounters> messageCounters = writerController.flushAll(); // blocks until complete
+
+      final long commitStartTime = System.currentTimeMillis();
+      consumerConnector.commitOffsets();
+
+      final long finishedTime = System.currentTimeMillis();
+      Map<String, MessageCounters> countsSinceLastCommit = new HashMap();
+      for (Map.Entry<String, MessageCounters> entry : messageCounters.entrySet()) {
+        countsSinceLastCommit.put(
+            entry.getKey(),
+            entry.getValue().difference(previousMessageCounters.get(entry.getKey()))
+        );
+      }
+
+      previousMessageCounters = messageCounters;
+
+      log.info(
+          "Flushed %s pending messages in %sms and committed offsets in %sms.",
+          countsSinceLastCommit.isEmpty() ? "0" : countsSinceLastCommit,
+          commitStartTime - flushStartTime,
+          finishedTime - commitStartTime
+      );
+    }
+    finally {
+      commitLock.writeLock().unlock();
+    }
+  }
+
+  private Runnable createCommitRunnable()
+  {
+    return new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        long lastFlushTime = System.currentTimeMillis();
+        try {
+          while (!Thread.currentThread().isInterrupted()) {
+            Thread.sleep(Math.max(commitMillis - (System.currentTimeMillis() - lastFlushTime), 0));
+            commit();
+            lastFlushTime = System.currentTimeMillis();
+          }
+        }
+        catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.info(e, "Commit thread interrupted.");
```

Comment:
The `e, ` here is going to make the log message scarier than it needs to be, & can probably be skipped.